src/main/java/upeldev/com/github/upel3/controllers/NewGradeController.java

```diff
@@ -63,17 +61,44 @@ public String createCourse(
 
         Activity currentActivity = activityService.findActivityById(activityId);
 
+
+
+        List<SubActivity> subActivities =  currentActivity.getSubActivity();
+
         try{
             if(!currentUser.getRoles().contains(Role.ADMIN) && !currentCourse.getLecturers().contains(currentUser)){
                 String errorMsg = "Nie masz wystarczających uprawnień, aby utworzyć kurs";
                 model.addAttribute("errorMsg", errorMsg);
-                return "error";
+                return "redirect:/error";
+            }
+
+            if(subActivities.size() != subActivityId.length){
+                String errorMsg = "Nie wypełniono wszystkich wymaganych pól";
+                model.addAttribute("errorMsg", errorMsg);
+                return "redirect:/error";
             }
-            Grade newGrade = new Grade(modifiedUser, currentActivity, gradeValue);
-            gradeService.save(newGrade);
 
-            List<Course> courses = currentUser.getCoursesLectured();
-            model.addAttribute("courses", courses);
+            if(gradeService.findGradeByCourseUserActivity(currentCourse, modifiedUser, currentActivity).size() == 0){
+                Grade newGrade = new Grade(modifiedUser, currentActivity);
+                newGrade.setDescription(description);
+                gradeService.save(newGrade);
+
+
+                for(int i = 0; i < subActivities.size(); i++){
+                    SubGrade subGrade = new SubGrade(subActivities.get(i), newGrade, subActivityId[i]);
+                    subGradeService.save(subGrade);
+                }
+            }
+            else{
+                Grade oldGrade = gradeService.findGradeByCourseUserActivity(currentCourse, modifiedUser, currentActivity).get(0);
+                oldGrade.setDescription(description);
+                for(int i = 0; i < subActivities.size(); i++){
+                    SubGrade subGrade = oldGrade.getSubGrade().get(i);
+                    subGrade.setValue(subActivityId[i]);
+                }
+                gradeService.save(oldGrade);
+
+            }
```

Comment:
I think this kind of logic should be a responsibility of service, not controller. Try to create functions for creating and saving a grade with a description or creating and saving a list of subgrades in appropriate services